### ISSUE TITLE
Inheritance of Behaviors

### ISSUE THREAD

**Issue by azaroth42** [IIIF/trc#13]
## Original Issue: 
https://github.com/IIIF/api/issues/1612 

## Background:

In 2.x we have the `viewingHint` property, that carries hints as to how the publisher would like the rendering client to process the information in the manifest. 

In 3.x, the `behavior` (alpha / beta) property replaces `viewingHint`, as the hint might not be about "viewing" per se. With A/V, it could be entirely audio. It also lends more weight to the recommendation, as opposed to just a "hint" that could be ignored completely with no undesirable effects.

## Issue:

When implementers came to try and understand the specification, there were questions (as raised in #1612) about when behaviors specified on classes higher in the hierarchy should be interpreted as being in effect for the child classes.  For example, if a Manifest is `paged`, does that imply that the Ranges are also `paged`? If a Collection is `unordered`, does that mean that all of its included Manifests are also `unordered`?

## Solution:

After much discussion on community technical calls and in smaller groups, the following inheritance rules were proposed as the simplest set that generated the most intuitive interactions: https://preview.iiif.io/api/image-prezi-rc2/api/presentation/3.0/#behavior

## Rationale:

Without these rules, clients could entirely legitimately either inherit unintended behaviors (making all manifests unordered) or not inherit intended behaviors (Ranges not inheriting paged from the manifest). If these inheritance rules are not part of the specification, then two implementations can have entirely different interpretations and be able to claim IIIF compliance. A viewer that shuffles the order of the canvases (pages) in manifests linked from any unordered collection would be a correct implementation, and this is clearly not desirable. Please note that implementation notes or recipes are intended to provide assistance in modeling and understanding, not to provide requirements, and should be able to be ignored completely when implementing.

It was considered too onerous for publishers to be required to be explicit about every behavior on every resource, and thus inheritance rules are required to avoid this situation.

One of the core principles of the version 3 specification shifts is to balance the focus on both publishing and consuming implementations.  Thus the additional clarity around what clients are expected to do, and the precision about the expected values for different properties. This issue is one of the client developer requested clarifications, and is important for ensuring consistency of the end user experience between different implementations.

**Comment by cubap** [comment]
well reasoned.

**Comment by aisaac** [comment]
I'm -1 for similar reasons as for #16 - in this specific case it's even worse because it's more complex, and in the spec it's the first thing readers see about behaviours

**Comment by aisaac** [comment]
I would add that at least if the spec would put the presentation of the rules *after* the presentation of the behaviours (and the example) at https://preview.iiif.io/api/image-prezi-rc2/api/presentation/3.0/#behavior, this would be a lesser evil in my view.

**Comment by tpendragon** [comment]
To be clear, inheritance here means that "the default comes from the parent unless one is supplied" - yes? Otherwise I would expect really complicated interaction scenarios between this and #16 (IE my MVW (a collection) is continuous and multi-part, but my volume is paged, is my volume now continuous, multi-part, and paged? I assume not.)

**Comment by mixterj** [comment]
I agree with @tpendragon and @aisaac. For issue 16 I think disjoint makes sense - you can not be both paged and continuous. But for issue 13 it seems to boil down to data producers not explicitly coding the expected behavior on all resources that Viewers need to have behaviors for. This is problematic for me on a few levels. From a RDF perspective the idea of inherent behavior from a parent object makes no sense (what is the parent object in a graph?). This is also problematic from a viewer standpoint when 'top level inheritance' objects are only referenced as a URI in the Manifest. The ease of navigating a JSON tree is gone, at least on initial Manifest load time, when you have to de-reference a Range to get the initial Range behavior (which I think is valid based on this documentation)

> Top level Ranges are embedded or externally referenced within the Manifest in a structures property. These top level Ranges then embed or reference other Ranges, Canvases or parts of Canvases in the items property. Each entry in the items property must be a JSON object, and it must have the id and type properties. If a top level Range needs to be dereferenced by the client, then it must not have the items property, such that clients are able to recognize that it should be retrieved in order to be processed.   

https://preview.iiif.io/api/image-prezi-rc2/api/presentation/3.0/#54-range

Finally, the idea that the Manifest would be too large if you explicitly coded all behaviors seems odd to me as well given how large Manifests can already get.

**Comment by azaroth42** [comment]
It's less the size of the manifest, but the annoyance to the producer of having to continually assert everything that could easily be inferred from the higher level resource. 

To continue the point about specificity and developer happiness, if a producer does not put all of the relevant behaviors on every resource, then what is the conforming client expected to do?  Throw an error and tell the end user that the Manifest is broken, or to make a best guess as to what was intended?

* If the former, then we need to take that back to the community and see how UV, Mirador and everyone feels about their users being told that a missing "paged" behavior on a Canvas will prevent their manuscript from being rendered at all.
* If the latter, then we then need to give guidance as to what that best guess is ... and then we need the inheritance + disjointness specifications ... which is what this is.

**Comment by azaroth42** [comment]
> I think it'd be easier to have a "default" behavior for each type in the specification

I disagree, not because this is a bad idea technically, but because the decision for which behaviors should be the defaults is an impossible one. For example, if we looked at the most commonly used behaviors, we would probably say that the default should be `paged`.  But then you're promoting the library use cases over top of archival, museum, STEM and other communities. So then you pick, say, `individuals` as the least likely to generate a bad experience and the majority of the data still has to be explicit for every single canvas of every book-like manifest that it's a page. So people ignore it, and clients try to be helpful, and we're back to needing clarity in the specifications.

**Comment by tpendragon** [comment]
> * If the former, then we need to take that back to the community and see how UV, Mirador and everyone feels about their users being told that a missing "paged" behavior on a Canvas will prevent their manuscript from being rendered at all.

FWIW, this is basically my experience now and I'm pretty cool with it ("what, that didn't work? Okay, let's look at examples, ah I need X, K, oh look it renders")

**Comment by aisaac** [comment]
For the record I've asked some Europeana developers their opinion. The one answer so far (I guess I can tell his name, but am reluctant to tag him along in this discussion) was a slight preference for having "less rules and restrictions". But he certainly did acknowledge the value of the other option for building robust viewers, and recognized all in all it was for him a matter of taste.
I'll keep you in the loop if others chime in.

**Comment by azaroth42** [comment]
### Issue 13 (Inheritance of Behaviors)
  +1: 21 [awead azaroth42 beaudet cubap emulatingkat gigamorph irv jonhartzler joshuago78 jronallo jtweed julsraemy jwd markpatton mattmcgrattan mcwhitaker regisrob rentonsa sredick tomcrane zimeon]
   0: 2 [jbhoward-dublin mejackreed]
  -1: 4 [aisaac kzhr mixterj tpendragon]

### Result: 21 / 27 = 0.78
Super majority is in favor, issue is approved
